server/src/charts/charts.service.ts

```diff
@@ -0,0 +1,100 @@
+import { forwardRef, Inject, Injectable } from '@nestjs/common'
+import { InjectRepository } from '@nestjs/typeorm'
+import { User } from 'src/auth/user.types'
+import { QueriesService } from 'src/queries/queries.service'
+import { Chart } from './chart.entity'
+import { ChartRepository } from './chart.repository'
+import { CreateChartDto, FilterChartDto, UpdateChartDto } from './charts.dto'
+
+@Injectable()
+export class ChartsService {
+  constructor(
+    @InjectRepository(ChartRepository)
+    private chartsRepository: ChartRepository,
+    @Inject(forwardRef(() => QueriesService))
+    private queriesService: QueriesService,
+  ) {}
+
+  /**
+   * Returns all the charts
+   *
+   * @param user
+   * @param query
+   */
+  getAllCharts(user: User, filterChartDto: FilterChartDto): Promise<Chart[]> {
+    const { query } = filterChartDto
+    if (query) {
+      return this.chartsRepository.find({
+        order: { createdAt: 'ASC' },
+        where: { uid: user.uid, query: { id: query } },
+      })
+    } else {
+      return this.chartsRepository.find({
+        order: { createdAt: 'ASC' },
+        where: { uid: user.uid },
+      })
+    }
+  }
+
+  /**
+   * Returns the chart data
+   *
+   * @param id
+   * @param user
+   */
+  async getChart(id: number, user: User): Promise<Chart> {
+    const chart = await this.chartsRepository.findOne({
+      id,
+      uid: user.uid,
+    })
+    return chart
+  }
+
+  /**
+   * Create a chart
+   *
+   * @param createChartDto
+   * @param user
+   */
+  async createChart(
+    createChartDto: CreateChartDto,
+    user: User,
+  ): Promise<Chart> {
+    const { query: queryId } = createChartDto
+    const query = await this.queriesService.getQuery(queryId, user)
+    return this.chartsRepository.save({
+      ...createChartDto,
+      query,
+      uid: user.uid,
+    })
+  }
+
+  /**
+   * Updates a particular chart
+   *
+   * @param id
+   * @param updateChartDto
+   * @param user
+   */
+  async updateChart(
+    id: number,
+    updateChartDto: UpdateChartDto,
+    user: User,
+  ): Promise<Chart> {
+    return this.chartsRepository
+      .update({ id, uid: user.uid }, updateChartDto)
+      .then(() => this.getChart(id, user))
+  }
+
+  /**
+   * Delete a particular chart
+   *
+   * @param id
+   * @param user
+   */
+  async deleteChart(id: number, user: User): Promise<Chart> {
+    const chart = await this.getChart(id, user)
+    await this.chartsRepository.remove([chart])
+    return { ...chart, id }
```

Comment:
The `id` value would be present in the `chart` object. Do we need to explicitly add it?

Reply:
```
{
  "success": true,
  "data": {
    "createdAt": "2021-01-14T02:33:49.102Z",
    "updatedAt": "2021-01-14T02:33:49.102Z",
    "name": "Revenue by country",
    "label": "product_name",
    "values": [
      "amount",
      "quantity"
    ],
    "type": "line",
    "config": null,
    "uid": "VnVekPavAHPzsuvF10zW0HAps1s1",
    "query": {
      "id": 16,
      "createdAt": "2020-12-26T04:16:50.553Z",
      "updatedAt": "2020-12-26T04:16:50.553Z",
      "name": "All Products",
      "query": "SELECT\n  *\nFROM\n  products;",
      "isSaved": true,
      "uid": "VnVekPavAHPzsuvF10zW0HAps1s1",
      "resource": {
        "id": 1,
        "createdAt": "2020-12-20T08:43:54.738Z",
        "updatedAt": "2020-12-20T08:43:54.738Z",
        "type": "postgres",
        "name": "Northwind",
        "host": "localhost",
        "port": 5432,
        "database": "northwind",
        "username": "admin",
        "password": "dopplr@123",
        "uid": "VnVekPavAHPzsuvF10zW0HAps1s1",
        "sslRequired": null,
        "selfCertificate": null,
        "clientKey": null,
        "clientCertificate": null
      }
    }
  }
}
```
Delete method doesn't send id in response. We have done the same thing in delete method for query